src/core/ComboBox/ComboBox.tsx

```diff
@@ -117,252 +123,317 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+        return;
+      }
+      onChangeProp.current?.(value);
+    }
+  }, [options, selectedIndex, valueProp]);
+
+  // Memoize the default (not selected/focused) version of menu items
+  const memoizedItems = React.useMemo(() => {
+    return options.map((option, index) => {
+      const { __originalIndex, ...rest } = option;
+
+      const customItem = itemRenderer
+        ? itemRenderer(option, {
+            isFocused: false,
+            isSelected: false,
+            index: __originalIndex,
+            id: option.id,
+          })
+        : null;
 
-    const _filteredOptions =
-      filterFunction?.(options, inputValue) ??
-      options.filter((option) =>
-        option.label.toLowerCase().includes(inputValue?.trim().toLowerCase()),
+      return customItem ? (
+        React.cloneElement(customItem, {
+          onClick: (e: unknown) => {
+            dispatch(['select', __originalIndex as number]);
+            customItem.props.onClick?.(e);
+          },
+          'data-iui-index': __originalIndex,
+        })
+      ) : (
+        <ComboBoxMenuItem
+          key={(__originalIndex as number) ?? index}
+          index={option.__originalIndex}
+          onClick={() => dispatch(['select', __originalIndex as number])}
+          {...rest}
+        >
+          {option.label}
+        </ComboBoxMenuItem>
       );
-    setFilteredOptions(_filteredOptions);
+    });
+  }, [itemRenderer, options]);
 
-    setFocusedIndex((previouslyFocusedIndex) => {
-      if (_filteredOptions.includes(options[previouslyFocusedIndex])) {
-        return previouslyFocusedIndex;
-      } else if (
-        _filteredOptions.find(({ value }) => value === selectedValue)
+  const getMenuItem = React.useCallback(
+    (option: ComboBoxOption<T>) => {
+      const { __originalIndex, ...rest } = option;
+
+      if (
+        selectedIndex !== __originalIndex &&
+        focusedIndex !== __originalIndex
       ) {
-        return options.findIndex(({ value }) => value === selectedValue);
-      } else {
-        return -1; // reset focus if previously focused or selected value is not in filtered list
+        return memoizedItems[option.__originalIndex];
       }
-    });
-  }, [inputValue, options, selectedValue, isOpen, filterFunction]);
-
-  const onKeyDown = React.useCallback(
-    (event: React.KeyboardEvent) => {
-      const focusableOptions = getFocusableElements(menuRef.current);
-      const focusedIndexInFilteredList = focusableOptions.findIndex(
-        ({ id }) =>
-          id === inputRef.current?.getAttribute('aria-activedescendant'),
+
+      const customItem = itemRenderer
+        ? itemRenderer(option, {
+            isFocused: focusedIndex === __originalIndex,
+            isSelected: selectedIndex === __originalIndex,
+            index: __originalIndex,
+            id: option.id,
+          })
+        : null;
+
+      return customItem ? (
+        React.cloneElement(customItem, {
+          onClick: (e: unknown) => {
```

Comment:
It mus be `li` element though, no?

Reply:
it should be `li` but it can be anything. we dont control it

Reply:
but wouldnt just `React.MouseEvent<HTMLElement>` work here?

Reply:
`MenuItem` doesn't expose the mouse event

Reply:
Huh? I dont understand. Click even is gonna be either `React.MouseEvent<HTMLElement>` or then will be no event at all (if `MenuItem` does not expose event at all).

Reply:
no, `MenuItem` (confusingly) passes the `value` of the item.

https://github.com/iTwin/iTwinUI-react/blob/4c9b4a8ae9973303c1ea9314de6958773308b8be/packages/iTwinUI-react/src/core/Menu/MenuItem.tsx#L28-L35
https://github.com/iTwin/iTwinUI-react/blob/4c9b4a8ae9973303c1ea9314de6958773308b8be/packages/iTwinUI-react/src/core/Menu/MenuItem.tsx#L146